homeassistant/components/home_plus_control/api.py

```diff
@@ -0,0 +1,352 @@
+            return
+
+        self.logger.debug(
+            "Legrand Home+ Control oauth token is no longer valid so refreshing."
+        )
+        new_token = await self.implementation.async_refresh_token(self.token)
+        self.token = new_token
+
+        self.hass.config_entries.async_update_entry(
+            self.config_entry, data={**self.config_entry.data, "token": new_token}
+        )
+
+    async def fetch_data(self):
+        """Get the latest data from the API.
+
+        Returns:
+            dict: Dictionary of switch modules in their updated state.
+        """
+        await self.async_handle_plant_data()
+        return await self.async_handle_module_status()
+
+    async def close_connection(self):
+        """Clean up the connection."""
+        await self.oauth_client.close()
+
+    def update_refresh_intervals(self):
+        """Update the time intervals that determine whether or not to call the API."""
+        if self.config_entry.options is not None:
+            for interval_name in DEFAULT_UPDATE_INTERVALS:
+                try:
+                    new_interval_value = int(self.config_entry.options[interval_name])
+                except KeyError:
+                    new_interval_value = DEFAULT_UPDATE_INTERVALS[interval_name]
+
+                self._refresh_intervals[interval_name] = new_interval_value
+                self.logger.debug(
+                    "Updated API refresh interval: %s: %s",
+                    interval_name,
+                    self._refresh_intervals[interval_name],
+                )
+
+    async def async_handle_plant_data(self):
+        """Recover the plant data for this particular user.
+
+        This will populate the "private" array of plants of this object and will return it.
+        It is expected that in most cases, there will only be one plant.
+
+        Returns:
+            dict: Dictionary of plants for this user - Keyed by the plant ID. Can be empty if no
+                  plants are retrieved.
+        """
+        # Attempt to recover the plant information from the Hass data object.
+        # If it is not there, then we request it from the API and add it.
+        # We also refresh from the API if the time has expired.
+        plant_info = self.hass.data[self._domain].get("plant_info")
+        if plant_info is None or self._should_check(
+            CONF_PLANT_UPDATE_INTERVAL,
+            self._refresh_intervals[CONF_PLANT_UPDATE_INTERVAL],
+        ):
+            try:
+                result = await self.get_request(PLANT_URL)  # Call the API
+                plant_info = await result.json()
+            except aiohttp.ClientError as err:
+                raise HomeAssistantError("Error retrieving plant information") from err
+
+            # If all goes well, we update the last check time
+            self._last_check[CONF_PLANT_UPDATE_INTERVAL] = time.monotonic()
+
+            self.hass.data[self._domain]["plant_info"] = plant_info
```

Comment:
It's weird that we have this side effect of storing data in hass.data, when the goal should be to have the coordinator manage the data returned from the api.